File: problems/0002-data-stream-transmission/solutions/python/solution.py
"""Problem: Data Stream Transmission.

Approach: exchange-argument ordering followed by sequence partition DP.
Time: O(n log n + k * n^2).
Space: O(n).
"""

from __future__ import annotations

import json
import sys
from typing import Any


Packet = tuple[int, float]


def solve(input_data: dict[str, Any]) -> float:
    """Return the minimum expected transmission time, rounded to six places."""
    packet_count = int(input_data["n"])
    checkpoint_count = int(input_data["k"])
    packets: list[Packet] = [
        (int(duration), float(success_probability))
        for duration, success_probability in input_data["packets"]
    ]

    def ordering_key(packet: Packet) -> float:
        duration, success_probability = packet
        failure_probability = 1.0 - success_probability
        return (
            duration / failure_probability
            if failure_probability > 0.0
            else float("inf")
        )

    packets.sort(key=ordering_key)

    success_prefix = [1.0] * (packet_count + 1)
    weighted_duration_prefix = [0.0] * (packet_count + 1)
    for index, (duration, success_probability) in enumerate(packets):
        success_prefix[index + 1] = success_prefix[index] * success_probability
        weighted_duration_prefix[index + 1] = (
            weighted_duration_prefix[index] + duration * success_prefix[index]
        )

    def segment_cost(start: int, end: int) -> float:
        return (
            weighted_duration_prefix[end + 1] - weighted_duration_prefix[start]
        ) / success_prefix[end + 1]

    costs = [segment_cost(0, end) for end in range(packet_count)]
    answer = costs[-1]

    maximum_segments = min(checkpoint_count + 1, packet_count)
    for segment_count in range(2, maximum_segments + 1):
        next_costs = [float("inf")] * packet_count
        for end in range(segment_count - 1, packet_count):
            inverse_success = 1.0 / success_prefix[end + 1]
            next_costs[end] = min(
                costs[split]
                + (
                    weighted_duration_prefix[end + 1]
                    - weighted_duration_prefix[split + 1]
                )
                * inverse_success
                for split in range(segment_count - 2, end)
            )
        costs = next_costs
        answer = min(answer, costs[-1])

    return round(answer, 6)


def main() -> None:
    json.dump(solve(json.load(sys.stdin)), sys.stdout)


if __name__ == "__main__":
    main()

File: problems/0002-data-stream-transmission/solutions/go/solution.go
// Problem: Data Stream Transmission
// Approach: Exchange-argument ordering followed by sequence partition DP.
// Time: O(n log n + k * n^2).
// Space: O(n).
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"sort"
)

type input struct {
	PacketCount     int          `json:"n"`
	CheckpointCount int          `json:"k"`
	Packets         [][2]float64 `json:"packets"`
}

type packet struct {
	duration           float64
	successProbability float64
}

func orderingKey(value packet) float64 {
	failureProbability := 1 - value.successProbability
	if failureProbability == 0 {
		return math.Inf(1)
	}
	return value.duration / failureProbability
}

func roundToSix(value float64) float64 {
	return math.RoundToEven(value*1_000_000) / 1_000_000
}

func solve(data input) float64 {
	packets := make([]packet, data.PacketCount)
	for index, values := range data.Packets {
		packets[index] = packet{
			duration:           values[0],
			successProbability: values[1],
		}
	}
	sort.SliceStable(packets, func(left, right int) bool {
		return orderingKey(packets[left]) < orderingKey(packets[right])
	})

	successPrefix := make([]float64, data.PacketCount+1)
	weightedDurationPrefix := make([]float64, data.PacketCount+1)
	successPrefix[0] = 1
	for index, value := range packets {
		successPrefix[index+1] = successPrefix[index] * value.successProbability
		weightedDurationPrefix[index+1] = weightedDurationPrefix[index] +
			value.duration*successPrefix[index]
	}

	segmentCost := func(start, end int) float64 {
		return (weightedDurationPrefix[end+1] - weightedDurationPrefix[start]) /
			successPrefix[end+1]
	}

	costs := make([]float64, data.PacketCount)
	for end := range costs {
		costs[end] = segmentCost(0, end)
	}
	answer := costs[data.PacketCount-1]
	maximumSegments := min(data.CheckpointCount+1, data.PacketCount)

	for segmentCount := 2; segmentCount <= maximumSegments; segmentCount++ {
		nextCosts := make([]float64, data.PacketCount)
		for index := range nextCosts {
			nextCosts[index] = math.Inf(1)
		}
		for end := segmentCount - 1; end < data.PacketCount; end++ {
			inverseSuccess := 1 / successPrefix[end+1]
			best := math.Inf(1)
			for split := segmentCount - 2; split < end; split++ {
				candidate := costs[split] +
					(weightedDurationPrefix[end+1]-weightedDurationPrefix[split+1])*
						inverseSuccess
				best = math.Min(best, candidate)
			}
			nextCosts[end] = best
		}
		costs = nextCosts
		answer = math.Min(answer, costs[data.PacketCount-1])
	}

	return roundToSix(answer)
}

func main() {
	var data input
	if err := json.NewDecoder(os.Stdin).Decode(&data); err != nil {
		fmt.Fprintln(os.Stderr, "invalid JSON input:", err)
		os.Exit(1)
	}
	if err := json.NewEncoder(os.Stdout).Encode(solve(data)); err != nil {
		fmt.Fprintln(os.Stderr, "failed to encode result:", err)
		os.Exit(1)
	}
}
